source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst

```diff
@@ -365,36 +365,60 @@ The hostname portion of each address MUST be normalized to lower-case.
+If the ``directConnection`` URI option is specified when a MongoClient is
+constructed, the TopologyType must be initialized based on the value of
+the ``directConnection`` option and the presence of the ``replicaSet`` option
+according to the following table:
+
++------------------+-----------------------+-----------------------+
+| directConnection | replicaSet present    | Initial TopologyType  |
++==================+=======================+=======================+
+| true             | no                    | Single                |
++------------------+-----------------------+-----------------------+
+| true             | yes                   | Single                |
++------------------+-----------------------+-----------------------+
+| false            | no                    | Unknown               |
++------------------+-----------------------+-----------------------+
+| false            | yes                   | ReplicaSetNoPrimary   |
++------------------+-----------------------+-----------------------+
+
+If the ``directConnection`` option is not specified, newly developed drivers
+MUST behave as if it was specified with the false value.
+
+Since changing the starting topology can reasonably be considered a
+backwards-breaking change, existing drivers SHOULD stage implementation
+according to semantic versioning guidelines. Specifically, support for the
+``directConnection`` URI option can be added in a minor release.
+In a subsequent major release, the default starting topology can be changed
+to Unknown. Drivers MUST document this in a prior minor release.
+
+Existing drivers MUST deprecate other URI options, if any, for controlling
+topology discovery or specifying the deployment topology. If such a legacy
+option is specified and the ``directConnection`` option is also
+specified, and the values of the two options are semantically different,
+the driver MUST report an error during URI option parsing.
+
+The API for initializing TopologyType using language-specific native options
+is not specified here. Drivers might already have a convention, e.g. a single
+seed means Single, a setName means ReplicaSetNoPrimary, and a list of seeds
+means Unknown. There are variations, however: In the Java driver a single seed
+means Single, but a **list** containing one seed means Unknown, so it can
+transition to replica-set monitoring if the seed is discovered to be a
+replica set member. In contrast, PyMongo requires a non-null setName in order
+to begin replica-set monitoring, regardless of the number of seeds.
+This spec does not cover language-specific native options that a driver may
+provide.
 
 Initial setName
 ~~~~~~~~~~~~~~~
 
-The user MUST be able to set the client's initial replica set name.
+It is allowed to use ``directConnection=true`` in conjunction with the
+``replicaSet`` URI option. The driver must connect in Single topology and
+verify that setName matches the specified name, as per
```

Comment:
Can you point me to the test that verifies that a non-matching replica set name with directConnection=true acts appropriately?  Something like

    mongodb://localhost/?directConnection=true&replicaSet=wrongName

If there isn't one, can it be added?

Reply:
I added one but looking at the resulting topology, it is not clear that such a topology will fail all operations. We handle this case [during server selection](https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/server_selector/selectable.rb#L218).